File: lib/models/notification_options.dart
import 'notification_channel_importance.dart';
import 'notification_priority.dart';
import 'notification_visibility.dart';

/// Notification options for Android platform.
class AndroidNotificationOptions {
  /// Constructs an instance of [AndroidNotificationOptions].
  AndroidNotificationOptions({
    @Deprecated('Use startService(serviceId) instead.') this.id,
    required this.channelId,
    required this.channelName,
    this.channelDescription,
    this.channelImportance = NotificationChannelImportance.LOW,
    this.priority = NotificationPriority.LOW,
    this.enableVibration = false,
    this.playSound = false,
    this.showWhen = false,
    this.showBadge = false,
    this.onlyAlertOnce = false,
    this.visibility = NotificationVisibility.VISIBILITY_PUBLIC,
  })  : assert(channelId.isNotEmpty),
        assert(channelName.isNotEmpty);

  /// Unique ID of the notification.
  final int? id;

  /// Unique ID of the notification channel.
  ///
  /// It is set only once for the first time on Android 8.0+.
  final String channelId;

  /// The name of the notification channel.
  ///
  /// It is set only once for the first time on Android 8.0+.
  final String channelName;

  /// The description of the notification channel.
  ///
  /// It is set only once for the first time on Android 8.0+.
  final String? channelDescription;

  /// The importance of the notification channel.
  /// The default is `NotificationChannelImportance.LOW`.
  ///
  /// It is set only once for the first time on Android 8.0+.
  final NotificationChannelImportance channelImportance;

  /// Priority of notifications for Android 7.1 and lower.
  /// The default is `NotificationPriority.LOW`.
  final NotificationPriority priority;

  /// Whether to enable vibration when creating notifications.
  /// The default is `false`.
  ///
  /// It is set only once for the first time on Android 8.0+.
  final bool enableVibration;

  /// Whether to play sound when creating notifications.
  /// The default is `false`.
  ///
  /// It is set only once for the first time on Android 8.0+.
  final bool playSound;

  /// Whether to show the timestamp when the notification was created in the content view.
  /// The default is `false`.
  final bool showWhen;

  /// Whether to show the badge near the app icon when service is started.
  /// The default is `false`.
  ///
  /// It is set only once for the first time on Android 8.0+.
  final bool showBadge;

  /// Whether to only alert once when the notification is created.
  /// The default is `false`.
  final bool onlyAlertOnce;

  /// Control the level of detail displayed in notifications on the lock screen.
  /// The default is `NotificationVisibility.VISIBILITY_PUBLIC`.
  final NotificationVisibility visibility;

  /// Returns the data fields of [AndroidNotificationOptions] in JSON format.
  Map<String, dynamic> toJson() {
    return {
      'notificationId': id,
      'notificationChannelId': channelId,
      'notificationChannelName': channelName,
      'notificationChannelDescription': channelDescription,
      'notificationChannelImportance': channelImportance.rawValue,
      'notificationPriority': priority.rawValue,
      'enableVibration': enableVibration,
      'playSound': playSound,
      'showWhen': showWhen,
      'showBadge': showBadge,
      'onlyAlertOnce': onlyAlertOnce,
      'visibility': visibility.rawValue,
    };
  }

  /// Creates a copy of the object replaced with new values.
  AndroidNotificationOptions copyWith({
    String? channelId,
    String? channelName,
    String? channelDescription,
    NotificationChannelImportance? channelImportance,
    NotificationPriority? priority,
    bool? enableVibration,
    bool? playSound,
    bool? showWhen,
    bool? showBadge,
    bool? onlyAlertOnce,
    NotificationVisibility? visibility,
  }) =>
      AndroidNotificationOptions(
        channelId: channelId ?? this.channelId,
        channelName: channelName ?? this.channelName,
        channelDescription: channelDescription ?? this.channelDescription,
        channelImportance: channelImportance ?? this.channelImportance,
        priority: priority ?? this.priority,
        enableVibration: enableVibration ?? this.enableVibration,
        playSound: playSound ?? this.playSound,
        showWhen: showWhen ?? this.showWhen,
        showBadge: showBadge ?? this.showBadge,
        onlyAlertOnce: onlyAlertOnce ?? this.onlyAlertOnce,
        visibility: visibility ?? this.visibility,
      );
}

/// Notification options for iOS platform.
class IOSNotificationOptions {
  /// Constructs an instance of [IOSNotificationOptions].
  const IOSNotificationOptions({
    this.showNotification = true,
    this.playSound = false,
  });

  /// Whether to show notifications.
  /// The default is `true`.
  final bool showNotification;

  /// Whether to play sound when creating notifications.
  /// The default is `false`.
  final bool playSound;

  /// Returns the data fields of [IOSNotificationOptions] in JSON format.
  Map<String, dynamic> toJson() {
    return {
      'showNotification': showNotification,
      'playSound': playSound,
    };
  }

  /// Creates a copy of the object replaced with new values.
  IOSNotificationOptions copyWith({
    bool? showNotification,
    bool? playSound,
  }) =>
      IOSNotificationOptions(
        showNotification: showNotification ?? this.showNotification,
        playSound: playSound ?? this.playSound,
      );
}

File: android/src/main/kotlin/com/pravera/flutter_foreground_task/models/NotificationOptions.kt
package com.pravera.flutter_foreground_task.models

import android.content.Context
import com.pravera.flutter_foreground_task.PreferencesKey as PrefsKey

data class NotificationOptions(
    val serviceId: Int,
    val channelId: String,
    val channelName: String,
    val channelDescription: String?,
    val channelImportance: Int,
    val priority: Int,
    val enableVibration: Boolean,
    val playSound: Boolean,
    val showWhen: Boolean,
    val showBadge: Boolean,
    val onlyAlertOnce: Boolean,
    val visibility: Int
) {
    companion object {
        fun getData(context: Context): NotificationOptions {
            val prefs = context.getSharedPreferences(
                PrefsKey.NOTIFICATION_OPTIONS_PREFS, Context.MODE_PRIVATE)

            val serviceId = prefs.getInt(PrefsKey.SERVICE_ID, prefs.getInt(PrefsKey.NOTIFICATION_ID, 1000))
            val channelId = prefs.getString(PrefsKey.NOTIFICATION_CHANNEL_ID, null) ?: "foreground_service"
            val channelName = prefs.getString(PrefsKey.NOTIFICATION_CHANNEL_NAME, null) ?: "Foreground Service"
            val channelDesc = prefs.getString(PrefsKey.NOTIFICATION_CHANNEL_DESC, null)
            val channelImportance = prefs.getInt(PrefsKey.NOTIFICATION_CHANNEL_IMPORTANCE, 2)
            val priority = prefs.getInt(PrefsKey.NOTIFICATION_PRIORITY, -1)
            val enableVibration = prefs.getBoolean(PrefsKey.ENABLE_VIBRATION, false)
            val playSound = prefs.getBoolean(PrefsKey.PLAY_SOUND, false)
            val showWhen = prefs.getBoolean(PrefsKey.SHOW_WHEN, false)
            val showBadge = prefs.getBoolean(PrefsKey.SHOW_BADGE, false)
            val onlyAlertOnce = prefs.getBoolean(PrefsKey.ONLY_ALERT_ONCE, false)
            val visibility = prefs.getInt(PrefsKey.VISIBILITY, 1)

            return NotificationOptions(
                serviceId = serviceId,
                channelId = channelId,
                channelName = channelName,
                channelDescription = channelDesc,
                channelImportance = channelImportance,
                priority = priority,
                enableVibration = enableVibration,
                playSound = playSound,
                showWhen = showWhen,
                showBadge = showBadge,
                onlyAlertOnce = onlyAlertOnce,
                visibility = visibility
            )
        }

        fun setData(context: Context, map: Map<*, *>?) {
            val prefs = context.getSharedPreferences(
                PrefsKey.NOTIFICATION_OPTIONS_PREFS, Context.MODE_PRIVATE)

            val serviceId = map?.get(PrefsKey.SERVICE_ID) as? Int
                ?: map?.get(PrefsKey.NOTIFICATION_ID) as? Int
                ?: 1000
            val channelId = map?.get(PrefsKey.NOTIFICATION_CHANNEL_ID) as? String
            val channelName = map?.get(PrefsKey.NOTIFICATION_CHANNEL_NAME) as? String
            val channelDesc = map?.get(PrefsKey.NOTIFICATION_CHANNEL_DESC) as? String
            val channelImportance = map?.get(PrefsKey.NOTIFICATION_CHANNEL_IMPORTANCE) as? Int ?: 2
            val priority = map?.get(PrefsKey.NOTIFICATION_PRIORITY) as? Int ?: -1
            val enableVibration = map?.get(PrefsKey.ENABLE_VIBRATION) as? Boolean ?: false
            val playSound = map?.get(PrefsKey.PLAY_SOUND) as? Boolean ?: false
            val showWhen = map?.get(PrefsKey.SHOW_WHEN) as? Boolean ?: false
            val showBadge = map?.get(PrefsKey.SHOW_BADGE) as? Boolean ?: false
            val onlyAlertOnce = map?.get(PrefsKey.ONLY_ALERT_ONCE) as? Boolean ?: false
            val visibility = map?.get(PrefsKey.VISIBILITY) as? Int ?: 1

            with(prefs.edit()) {
                putInt(PrefsKey.SERVICE_ID, serviceId)
                putString(PrefsKey.NOTIFICATION_CHANNEL_ID, channelId)
                putString(PrefsKey.NOTIFICATION_CHANNEL_NAME, channelName)
                putString(PrefsKey.NOTIFICATION_CHANNEL_DESC, channelDesc)
                putInt(PrefsKey.NOTIFICATION_CHANNEL_IMPORTANCE, channelImportance)
                putInt(PrefsKey.NOTIFICATION_PRIORITY, priority)
                putBoolean(PrefsKey.ENABLE_VIBRATION, enableVibration)
                putBoolean(PrefsKey.PLAY_SOUND, playSound)
                putBoolean(PrefsKey.SHOW_WHEN, showWhen)
                putBoolean(PrefsKey.SHOW_BADGE, showBadge)
                putBoolean(PrefsKey.ONLY_ALERT_ONCE, onlyAlertOnce)
                putInt(PrefsKey.VISIBILITY, visibility)
                commit()
            }
        }

        fun clearData(context: Context) {
            val prefs = context.getSharedPreferences(
                PrefsKey.NOTIFICATION_OPTIONS_PREFS, Context.MODE_PRIVATE)

            with(prefs.edit()) {
                clear()
                commit()
            }
        }
    }
}
